### PR TITLE
Add string repeat builtin function

### PR DESCRIPTION
**Pull request by Ivan-Z**
Closes: #3393

Signed-off-by: Ivan Zdane <mail@ivanzdane.com>

### CODE REVIEW

File: ast/builtins.go [hunk 1]
```diff
@@ -1034,6 +1035,18 @@ var Sprintf = &Builtin{
 	),
 }
 
+// Repeat returns the string s, repeated count number of times.
+var Repeat = &Builtin{
+	Name: "repeat",
```
Comment (srenatus) [inline ast/builtins.go:1040]: Let's make this `strings.repeat`, what do you think?

File: docs/content/policy-reference.md [hunk 2]
```diff
@@ -359,6 +359,7 @@ complex types.
 | <span class="opa-keep-it-together">``output := trim_suffix(string, suffix)``</span> | ``output`` is a ``string`` representing ``string`` with trailing instance of ``suffix`` removed. If ``string`` doesn't end with suffix, ``string`` is returned unchanged.| ✅ |
 | <span class="opa-keep-it-together">``output := trim_space(string)``</span> | ``output`` is a ``string`` representing ``string`` with all leading and trailing white space removed.| ✅ |
 | <span class="opa-keep-it-together">``output := upper(string)``</span> | ``output`` is ``string`` after converting to upper case | ✅ |
+| <span class="opa-keep-it-together">``output := repeat(string, count)``</span> | ``string`` is repeated count number of times | ✅ |
```
Comment (srenatus) [inline docs/content/policy-reference.md:362]:
```suggestion
| <span class="opa-keep-it-together">``output := repeat(string, count)``</span> | ``string`` is repeated ``count`` number of times | ✅ |
```

File: topdown/strings.go [hunk 3]
```diff
@@ -384,6 +384,25 @@ func builtinSprintf(a, b ast.Value) (ast.Value, error) {
...
+		return err
+	}
+
+	count, err := builtins.IntOperand(operands[1].Value, 2)
+	if err != nil {
+		return err
+	}
+
+	if count < 0 {
+		return builtins.NewOperandErr(2, "count must be greater than zero")
```
Comment (srenatus) [inline topdown/strings.go:399]: 💭 what's up with `repeat("hi", 0)`? It seems like it's valid, but it's also not _greater than zero_ 😉

**Review by srenatus — COMMENTED** [review]
This is a fine contribution, thanks for working on it. I've got a few smaller issues, mostly re: `repeat` vs `strings.repeat` (I'm afraid changing the name means touching all files again 😉)

Also, you don't need to commit any changes to `internal/compiler/opa/wasm` for this PR -- those would be auto-generated post-merge if they are required; but for this, they aren't even going to change.

File: ast/builtins.go (hunk 1 above)
Comment (tsandall) [reply ast/builtins.go:1040]: 👍 let's use `strings.repeat`. The bare function names are just here for legacy reasons. All new builtins should have a prefix.

File: topdown/strings.go (hunk 3 above)
Comment (Ivan-Z) [reply topdown/strings.go:399]: Gonna make it strictly greater than zero,  `repeat("hi", 0)` seems like a weird way of defining an empty string.

**Comment by Ivan-Z** [comment]
@srenatus Could you take another look please?

File: test/cases/testdata/strings/test-strings-0924.yaml [hunk 4]
```diff
@@ -0,0 +1,12 @@
+cases:
+- note: 'strings/repeat_test'
+  modules:
+  - |
+    package generated
```
Comment (ashutosh-narkar) [inline test/cases/testdata/strings/test-strings-0924.yaml:5]: Nit: `generated` -> `test`. Also it would be helpful to add a test case with bad operands.

**Review by ashutosh-narkar — COMMENTED** [review]
@Ivan-Z changes look good. Few minor comments.

File: docs/content/policy-reference.md [hunk 5]
```diff
@@ -359,6 +359,7 @@ complex types.
 | <span class="opa-keep-it-together">``output := trim_suffix(string, suffix)``</span> | ``output`` is a ``string`` representing ``string`` with trailing instance of ``suffix`` removed. If ``string`` doesn't end with suffix, ``string`` is returned unchanged.| ✅ |
 | <span class="opa-keep-it-together">``output := trim_space(string)``</span> | ``output`` is a ``string`` representing ``string`` with all leading and trailing white space removed.| ✅ |
 | <span class="opa-keep-it-together">``output := upper(string)``</span> | ``output`` is ``string`` after converting to upper case | ✅ |
+| <span class="opa-keep-it-together">``output := strings.repeat(string, count)``</span> | ``string`` is repeated ``count`` number of times | ``SDK-dependent`` |
```
[the remark this answers is not in the document]
Comment (Ivan-Z) [reply docs/content/policy-reference.md]: updated

File: test/cases/testdata/strings/test-strings-0924.yaml (hunk 4 above)
Comment (Ivan-Z) [reply test/cases/testdata/strings/test-strings-0924.yaml]: updated and added a test

File: test/cases/testdata/strings/test-strings-0925.yaml [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+  - |
+    package test
+
+    p = x {
+      x := strings.repeat("hi", -1)
+    }
+  query: data.test.p = x
+  note: 'count must be greater than zero'
+  want_result:
+  - x: ""
```
Comment (ashutosh-narkar) [inline test/cases/testdata/strings/test-strings-0925.yaml:13]: Something like [this](https://github.com/open-policy-agent/opa/blob/main/test/cases/testdata/replacen/test-replacen-bad-operands.yaml) test would be helpful.

**Comment by anderseknert** [comment]
Since we've had some problems with unicode in our string functions in the past, it would be great to also see a test like:

```rego
strings.repeat("☀️", 3) == "☀️☀️☀️"
```

**Comment by tsandall** [comment]
hey @Ivan-Z let us know if you'll be able to come back to this sometime soon.

**Comment by tsandall** [comment]
Closing this PR for now. We can come back to this in the future.

**Pull request closed without merging**